akka-persistence-query/src/main/scala/akka/persistence/query/Offset.scala

```diff
@@ -117,6 +117,61 @@ final class TimestampOffset private (val timestamp: Instant, val readTimestamp:
     s"TimestampOffset($timestamp, $readTimestamp, $seen)"
 }
 
+/**
+ * Timestamp-based offset by slice.
+ *
+ * API May Change
+ */
+object TimestampOffsetBySlice {
+  val empty: TimestampOffsetBySlice = new TimestampOffsetBySlice(Map.empty)
+
+  def apply(offsets: Map[Int, TimestampOffset]): TimestampOffsetBySlice =
+    new TimestampOffsetBySlice(offsets)
+
+  /** Java API */
+  def create(offsets: java.util.Map[java.lang.Integer, TimestampOffset]): TimestampOffsetBySlice = {
+    import akka.util.ccompat.JavaConverters._
+    new TimestampOffsetBySlice(offsets.asScala.toMap.map { case (slice, offset) => slice.intValue -> offset })
+  }
+
+  def unapply(timestampOffsetBySlice: TimestampOffsetBySlice): Option[Map[Int, TimestampOffset]] =
+    Some(timestampOffsetBySlice.offsets)
+}
+
+/**
+ * Timestamp-based offset by slice.
+ *
+ * API May Change
+ *
+ * @param offsets
+ *   Map of TimestampOffset by slice
+ */
+final class TimestampOffsetBySlice private (val offsets: Map[Int, TimestampOffset]) extends Offset {
```

Comment:
Ok. Should only be used on init I guess, not in envelopes, but makes sense to have coverage. So that's a string-based serialization format? I'll take a look. Maybe double the timestamp offset separator with `;;`.

Reply:
Added serialization. Uses `;;` for separator between entries, and `;` between each slice and its encoded timestamp offset.